x/nexus/keeper/transfer.go

```diff
@@ -60,8 +64,55 @@ func (k Keeper) getTransferFee(ctx sdk.Context) (fee exported.TransferFee) {
 	return fee
 }
 
+// computeChainSurcharge computes the surcharge fee over the base fee
+//
+// max_surcharge = chain.max_fee - chain.min_fee
+//
+// surcharge = min(max_surcharge, chain.fee_rate * (amount - baseFee))
+func computeChainSurcharge(feeInfo exported.FeeInfo, baseFee sdk.Int, amount sdk.Int) sdk.Int {
+	maxSurcharge := sdk.Int(feeInfo.MaxFee.Sub(feeInfo.MinFee))
+
+	remaining := sdk.ZeroInt()
+	if baseFee.LT(amount) {
+		remaining = amount.Sub(baseFee)
+	}
+
+	surcharge := sdk.NewDecFromInt(remaining).Mul(feeInfo.FeeRate).TruncateInt()
+	surcharge = sdk.MinInt(maxSurcharge, surcharge)
+
+	return surcharge
+}
+
+// ComputeTransferFee computes the fee for a cross-chain transfer.
+// If fee_info is not set for an asset on a chain, default of zero is used
+//
+// base_fee = source_chain.min_fee + destination_chain.min_fee
+//
+// transfer_fee = baseFee + source_chain_surcharge + destination_chain_surcharge
+//
+// INVARIANT: source_chain.min_fee + destination_chain.min_fee <= transfer_fee <= source_chain.max_fee + destination_chain.max_fee
+func (k Keeper) ComputeTransferFee(ctx sdk.Context, sourceChain exported.Chain, destinationChain exported.Chain, asset sdk.Coin) (sdk.Coin, exported.FeeInfo) {
```

Comment:
I don't understand why you do such a complicated calculation. I'd prefer instead (pseudo code):
```
minFee = sourceMin + destinationMin
rateFee = amount * (sourceRate + destinationRate)
maxFee = sourceMax + destinationMax

if minFee > rateFee{
return minFee
}

if maxFee < rateFee{
return maxFee
}

return rateFee
```

Reply:
Couple of points, but I have updated it to keep it simple for now:
- Applying rate to the whole amount means we do `max(min_fee, x% of amount)` vs `min_fee + x% of remaining`, but users probably expect the latter I guess.
- `min(rateFee, maxFee)` can be different than `min(sourceMax, amount * sourceRate) + min(destinationMax, amount * destinationRate)`.

Reply:
Let's not assume what the users expect. I do agree with @cgorenflo here that simply applying a parent on the entire balance makes a lot simpler formula than applying on the reminder.

Reply:
Chris's suggested calc makes sense.

Reply:
I meant to say they expect the former, haha.